lib/internal/urlCache.js

```diff
@@ -0,0 +1,84 @@
+"use strict";
+
+
+
+function urlCache()
+{
+	var thisObj = this;
+	this.cache = {};
+	this.retrievalStore = {};
+}
+
+
+
+urlCache.prototype.clear = function()
+{
+	this.cache = {};
+	this.retrievalStore = {};
+}
+
+
+
+urlCache.prototype.store = function(url, executionFn)
+{
+	this.cache[url] = executionFn;
+
+	if (Array.isArray(this.retrievalStore[url]) === true)
+	{
+		for (var i = 0; i < this.retrievalStore[url].length; i++)
```

Comment:
We should cache the length too, even though it's a micro-optimization:
```js
for (i=0; i<numUrls; i++)
```

and put all the `var` declarations at the very top of the function (to visualize "hoisting"):
```js
var i,numUrls,urls;
```